pkg/iac/terraform/block.go

```diff
@@ -576,3 +596,122 @@ func (b *Block) IsNil() bool {
 func (b *Block) IsNotNil() bool {
 	return !b.IsNil()
 }
+
+func (b *Block) ExpandBlock() error {
+	var (
+		expanded []*Block
+		errs     []error
+	)
+
+	for _, child := range b.childBlocks {
+		if child.Type() == "dynamic" {
+			blocks, err := child.expandDynamic()
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+			expanded = append(expanded, blocks...)
+		}
+	}
+
+	for _, block := range expanded {
+		b.injectBlock(block)
+	}
+
+	return errors.Join(errs...)
+}
+
+func (b *Block) expandDynamic() ([]*Block, error) {
+	if b.IsExpanded() || b.Type() != "dynamic" {
+		return nil, nil
+	}
+
+	realBlockType := b.TypeLabel()
+	if realBlockType == "" {
+		return nil, errors.New("dynamic block must have 1 label")
+	}
+
+	forEachVal, err := b.validateForEach()
+	if err != nil {
+		return nil, fmt.Errorf("invalid for-each in %s block: %w", b.FullLocalName(), err)
+	}
+
+	var (
+		expanded []*Block
+		errs     []error
+	)
+
+	forEachVal.ForEachElement(func(key, val cty.Value) (stop bool) {
+		if val.IsNull() {
+			return
+		}
+
+		iteratorName, err := b.iteratorName(realBlockType)
+		if err != nil {
+			errs = append(errs, err)
+			return
+		}
+
+		forEachCtx := b.childContext()
+		obj := cty.ObjectVal(map[string]cty.Value{
+			"key":   key,
+			"value": val,
+		})
+		forEachCtx.Set(obj, iteratorName)
+
+		if content := b.GetBlock("content"); content != nil {
+			inherited := content.inherit(forEachCtx)
+			inherited.hclBlock.Labels = []string{}
+			inherited.hclBlock.Type = realBlockType
+			if err := inherited.ExpandBlock(); err != nil {
+				errs = append(errs, err)
+				return
+			}
+			expanded = append(expanded, inherited)
+		}
+		return
+	})
+
+	if len(expanded) > 0 {
+		b.markExpanded()
+	}
+
+	return expanded, errors.Join(errs...)
+}
+
+func (b *Block) validateForEach() (cty.Value, error) {
```

Comment:
Can we add tests for the error paths in this func?

Reply:
[This](https://github.com/aquasecurity/trivy/pull/7612/files#diff-9a5bef5990c07910f49b85bf9dcbf4f2c28cd86eadd9d626730c8c35cf23d745L1369) test covers those cases

Reply:
I added one more case here 2becf4c67ee22a94ebd58a6c5c05db34de16caf9